cirq-core/cirq/transformers/eject_z_test.py

```diff
@@ -39,6 +42,33 @@ def assert_optimizes(
     circuit = cirq.eject_z(before, eject_parameterized=eject_parameterized, context=context)
     cirq.testing.assert_same_circuits(circuit, expected)
 
+    # Nested sub-circuits should also get optimized.
+    q = before.all_qubits()
+    c_nested = cirq.Circuit(
+        [(cirq.Z ** 0.5).on_each(*q), (cirq.Y ** 0.25).on_each(*q)],
+        cirq.Moment(cirq.CircuitOperation(before.freeze()).repeat(2).with_tags("ignore")),
+        [(cirq.Z ** 0.5).on_each(*q), (cirq.Y ** 0.25).on_each(*q)],
+        cirq.Moment(cirq.CircuitOperation(before.freeze()).repeat(3).with_tags("preserve_tag")),
+    )
+    c_expected = cirq.Circuit(
+        cirq.PhasedXPowGate(phase_exponent=0, exponent=0.25).on_each(*q),
+        (cirq.Z ** 0.5).on_each(*q),
+        cirq.Moment(cirq.CircuitOperation(before.freeze()).repeat(2).with_tags("ignore")),
+        cirq.PhasedXPowGate(phase_exponent=0, exponent=0.25).on_each(*q),
+        (cirq.Z ** 0.5).on_each(*q),
+        cirq.Moment(cirq.CircuitOperation(expected.freeze()).repeat(3).with_tags("preserve_tag")),
+    )
+    if context is None:
+        context = cirq.TransformerContext(tags_to_ignore=("ignore",), deep=True)
+    else:
+        context = dataclasses.replace(
+            context, tags_to_ignore=context.tags_to_ignore + ("ignore",), deep=True
+        )
+    c_nested = cirq.eject_z(c_nested, context=context, eject_parameterized=eject_parameterized)
+    cirq.testing.assert_same_circuits(c_nested, c_expected)
+    c_nested = cirq.eject_z(c_nested, context=context, eject_parameterized=eject_parameterized)
+    cirq.testing.assert_same_circuits(c_nested, c_expected)
```

Comment:
Can we move this to its own test instead of sticking it inside `assert_optimizes` ?

Reply:
The idea of putting it inside assert_optimizes is to test the nested case for all non-nested cases since bugs can easily creep in while testing only a few cases for nested sub-circuits (as seen in recent PRs for transformer primitives). Added a nested case to assert_optimizes ensures a broader coverage.

Reply:
SG